homeassistant/components/switcher_kis/config_flow.py

```diff
@@ -0,0 +1,49 @@
+"""Config flow for Switcher integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from homeassistant import config_entries
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers.typing import ConfigType
+
+from .const import DATA_DISCOVERY, DOMAIN
+from .utils import async_discover_devices
+
+
+class SwitcherFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle Switcher config flow."""
+
+    async def async_step_import(self, import_config: ConfigType) -> FlowResult:
+        """Handle a flow initiated by import."""
+        if self._async_current_entries(True):
+            return self.async_abort(reason="single_instance_allowed")
+
+        return self.async_create_entry(title="Switcher", data={})
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the start of the config flow."""
+        if self._async_current_entries(True):
+            return self.async_abort(reason="single_instance_allowed")
+
+        self.hass.data.setdefault(DOMAIN, {})
+        if DATA_DISCOVERY not in self.hass.data[DOMAIN]:
+            self.hass.data[DOMAIN][DATA_DISCOVERY] = self.hass.async_create_task(
```

Comment:
The config flow should not set objects or access objects in hass.data. The config flow should be isolated from the integration set up.

Reply:
I discussed this before implementation with @balloob,
I will probably move this to `register_discovery_flow` which remove all the config flow steps, but the discovery function will still need a way to pass the discovered devices to `async_setup_entry` else the user experience will not be good.

In current implementation:
1. User add the integration a background discovery is collecting devices
2. User approve adding the integration - discovery task is waited to be finished: 6 seconds - time waited in step 1, in most cases it will be 2-3 seconds.
3. Entries are displayed to assign area.

With changing to `register_discovery_flow` step 1 will not happen in the background so user will need to wait 6 seconds after approving adding the integration (less optimal than first option, but still acceptable)

If I can't store anything to be passed to the `async_setup_entry` in some way the user experience will be:
1. Upon approval user waits 6 seconds.
2. After approval user needs to wait another 6 seconds for the entries to be displayed

This will give a bad user experience - waiting 12 seconds instead of 4-6  as today, I understand that data is stored in `hass.data` which might not be used if the user cancel the flow, but this is a very rare use case, and in this case some very small data will be left in `hass.data`.

Reply:
I don't think it's that bad since it's only during setup. It's not something the user will do often at all.